CRM/Contribute/BAO/Contribution.php

```diff
@@ -3182,13 +3182,13 @@ public function _gatherMessageValues($input, &$values, $ids = []) {
             ]);
             $values['priceSetID'] = $priceSet['id'];
           }
+          $memberships = self::getRelatedMemberships($this->id);
           foreach ($lineItems as &$eachItem) {
-            if (isset($this->_relatedObjects['membership'])
-              && is_array($this->_relatedObjects['membership'])
-              && array_key_exists($eachItem['membership_type_id'], $this->_relatedObjects['membership'])) {
-              $eachItem['join_date'] = CRM_Utils_Date::customFormat($this->_relatedObjects['membership'][$eachItem['membership_type_id']]->join_date);
-              $eachItem['start_date'] = CRM_Utils_Date::customFormat($this->_relatedObjects['membership'][$eachItem['membership_type_id']]->start_date);
-              $eachItem['end_date'] = CRM_Utils_Date::customFormat($this->_relatedObjects['membership'][$eachItem['membership_type_id']]->end_date);
+            if (!empty($eachItem['membership_type_id'])) {
+              $membership = $memberships[$eachItem['entity_id']];
```

Comment:
@eileenmcnaughton we can have both PR (my with modification), until we stop using deprecated function. 

> @sunilpawar the test fail is a bug with the test - which will fix - but if we keep the scope for this PR to only addressing transition to completed - does this work as far as you are concerned
> 
> (I'll look at cancel separately)

Since you are using entity id (membership id) to get details from item array, i can modify my PR to use membership id as key to  `$this->_relatedObjects['membership']` instead of combination member id + membership type.

So i just need modification at one place 

```diff
   -     $this->_relatedObjects['membership'][$membership->membership_type_id] = $membership;
   +     $this->_relatedObjects['membership'][$membership->id] = $membership;
```

Reply:
@sunilpawar I'd rather take the extra time to get rid of the deprecated function call & be done with it....